File: AlexAndCircularTour.py
def is_cycle(node, vis, rec_stack, st, adj):
    vis[node] = True
    rec_stack[node] = True
    st.append(node)

    for neighbor in adj[node]:
        if not vis[neighbor]:
            if is_cycle(neighbor, vis, rec_stack, st, adj):
                return True
        elif rec_stack[neighbor]:
            st.append(neighbor)
            return True

    rec_stack[node] = False
    st.pop()
    return False

def extract(st):
    node = st[-1]
    cycle = []
    cycle.append(node)
    st.pop()

    while st:
        top = st[-1]
        cycle.append(top)
        if top == node:
            break
        st.pop()
    
    return cycle

def round_trip(n, adj):
    vis = [False] * (n + 1)
    rec_stack = [False] * (n + 1)
    st = []

    for i in range(1, n + 1):
        if not vis[i]:
            if is_cycle(i, vis, rec_stack, st, adj):
                return extract(st)
    
    return []

def main():
    import sys
    input = sys.stdin.read
    data = list(map(int, input().split()))

    idx = 0
    n = data[idx]
    idx += 1
    m = data[idx]
    idx += 1

    adj = [[] for _ in range(n + 1)]

    for _ in range(m):
        a = data[idx]
        idx += 1
        b = data[idx]
        idx += 1
        adj[a].append(b)

    cycle = round_trip(n, adj)

    if not cycle:
        print("NOT POSSIBLE")
    else:
        cycle.reverse()
        print(len(cycle))
        print(" ".join(map(str, cycle)))

if __name__ == "__main__":
    main()

File: AlexAndCircularTour.java
import java.util.*;

public class Main {
    static boolean isCycle(int node, boolean[] vis, boolean[] recStack, Stack<Integer> st, List<Integer>[] adj) {
        vis[node] = true;
        recStack[node] = true;
        st.push(node);

        for (int neighbor : adj[node]) {
            if (!vis[neighbor]) {
                if (isCycle(neighbor, vis, recStack, st, adj)) {
                    return true;
                }
            } else if (recStack[neighbor]) {
                st.push(neighbor);
                return true;
            }
        }

        recStack[node] = false;
        st.pop();
        return false;
    }

    static List<Integer> extract(Stack<Integer> st) {
        int node = st.peek();
        List<Integer> cycle = new ArrayList<>();
        cycle.add(node);
        st.pop();

        while (!st.isEmpty()) {
            int top = st.peek();
            cycle.add(top);
            if (top == node) break;
            st.pop();
        }
        return cycle;
    }

    static List<Integer> roundTrip(int n, List<Integer>[] adj) {
        boolean[] vis = new boolean[n + 1];
        boolean[] recStack = new boolean[n + 1];
        Stack<Integer> st = new Stack<>();

        for (int i = 1; i <= n; i++) {
            if (!vis[i]) {
                if (isCycle(i, vis, recStack, st, adj)) {
                    return extract(st);
                }
            }
        }
        return Collections.emptyList();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();

        List<Integer>[] adj = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) {
            adj[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            int a = sc.nextInt();
            int b = sc.nextInt();
            adj[a].add(b);
        }

        List<Integer> cycle = roundTrip(n, adj);

        if (cycle.isEmpty()) {
            System.out.println("NOT POSSIBLE");
        } else {
            Collections.reverse(cycle);
            System.out.println(cycle.size());
            for (int node : cycle) {
                System.out.print(node + " ");
            }
        }
    }
}
